### PR TITLE
Add Playwright

### PR DESCRIPTION
**Pull request by kumaraditya303**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by kumaraditya303** [comment]
@conda-forge-admin ping team

**Comment by kumaraditya303** [comment]
@conda-forge/help-python The windows build is failing because of an assertion error

**Comment by kumaraditya303** [comment]
@conda-forge/help-python The linux build is failing as even after settings `binary_relocation: False` it is trying to locate the shared libraries

**Comment by mxschmitt** [comment]
I want to be added as a maintainer.

**Comment by kumaraditya303** [comment]
ping @conda-forge/staged-recipes this is ready for review.

**Comment by kumaraditya303** [comment]
@conda-forge-admin ping team

**Comment by kumaraditya303** [comment]
@conda-forge/staged-recipes ready for review

File: recipes/playwright/meta.yaml
Comment (fcollonval) [inline recipes/playwright/meta.yaml:9]: I don't think your recipe will be accepted without using a legit source, like non-fork github - although I know the latest release is not working. But you should at least use the `version` variable to enforce homogeneous version.

```suggestion
  url: https://github.com/microsoft/playwright-python/archive/v{{ version }}.tar.gz
```

File: recipes/playwright/meta.yaml
Comment (fcollonval) [inline recipes/playwright/meta.yaml:42]: You must add constraints identical to the python package:

```
        "websockets>=8.1",
        "greenlet>=0.4",
        "pyee>=8.0.1",
        "typing-extensions;python_version<='3.8'",
```

See https://github.com/microsoft/playwright-python/blob/master/setup.py#L143

File: recipes/playwright/meta.yaml
Comment (fcollonval) [inline recipes/playwright/meta.yaml:50]: It is a good practice to ensure the conda package is compliant with pip

```suggestion
  requires:
    - pip
  imports:
    - playwright
    - playwright.sync_api
    - playwright.async_api
  commands:
    - pip check
    - playwright --help
```

File: recipes/playwright/meta.yaml
Comment (fcollonval) [inline recipes/playwright/meta.yaml:56]:
```suggestion
  license_file: LICENSE
```

As the source code comes from a GitHub archive, the _official_ license file is contained within it. So you should remove the license file in this recipe a link to the `LICENSE` file from the GitHub archive.

File: recipes/playwright/meta.yaml
Comment (fcollonval) [inline recipes/playwright/meta.yaml:36]: You don't need to set a `build` environment as you are not using a compiler. And the `host` environment should contain only the setup requirements.

```suggestion
  host:
    - python
    - pip
    - curl
    - setuptools_scm
```

**Review by fcollonval — COMMENTED** [review]
Hey @kumaraditya303 thank you for your awesome work on this.

I don't have rights to approve this PR. But I do have some recommendation to help it be accepted.

File: recipes/playwright/meta.yaml
Comment (kumaraditya303) [reply recipes/playwright/meta.yaml:9]: @fcollonval I know that but that wouldn't work because the last released version of it at Microsoft repo does not contains many fixes for installation directly from repo See https://github.com/microsoft/playwright-python/pull/537 I am waiting for a new release and then I would change it to official, Thanks for your suggestion

**Comment by kumaraditya303** [comment]
@fcollonval Do you know how to upgrade the glibc version ?https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=308581&view=logs&j=6f142865-96c3-535c-b7ea-873d86b887bd&t=22b0682d-ab9e-55d7-9c79-49f3c3ba4823&l=1890

File: recipes/playwright/meta.yaml
Comment (kumaraditya303) [reply recipes/playwright/meta.yaml]: Thanks

**Comment by fcollonval** [comment]
> @fcollonval Do you know how to upgrade the glibc version ?https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=308581&view=logs&j=6f142865-96c3-535c-b7ea-873d86b887bd&t=22b0682d-ab9e-55d7-9c79-49f3c3ba4823&l=1890

The major trouble is that conda is using its own build chain: https://conda-forge.org/docs/maintainer/infrastructure.html#compilers-and-runtimes

And the playwright driver is not compiled with that build chain (this is the [hard point](https://github.com/microsoft/playwright-python/issues/257#issuecomment-769932935) mentioned by @jtpio). My experience in that area is light. But I guess the driver should be created using that build chain.

**Comment by kumaraditya303** [comment]
I am thinking to just leave it because `conda-forge` uses an old distro which is centos 7 I think in this case which does not has recent glibc, and users are to be using in on ubuntu distro

**Comment by fcollonval** [comment]
Maybe there is a possible workaround as the trouble is raised by the node binary package within the driver. If it can be removed and replace to use the node conda package that may solve the issue.

**Comment by kumaraditya303** [comment]
That would be unnecessary as no one is going to use that old distro and most importantly `playwright` only supports ubuntu so it is out of scope

**Comment by fcollonval** [comment]
> That would be unnecessary as no one is going to use that old distro 

It is definitely still around.

> most importantly `playwright` only supports ubuntu so it is out of scope

I have never seen that requirement - from the documentation, it is only required for Firefox.

**Comment by kumaraditya303** [comment]
Well users could use old distro but that would require them to install the browser and their dependencies themselves and also if we want to support a distro then it should support all the browsers not just chromium.

**Comment by mxschmitt** [comment]
Sorry for jumping in. Just wanted to clarify that like kumaraditya303 said, only Ubuntu 18 and 20 are currently supported from the Playwright side. 👍

Let me add that to the docs next week.

**Comment by kumaraditya303** [comment]
Clarifying a bit more, `playwright` currently only supports Ubuntu 18 and Ubuntu 20 which are LTS releases of ubuntu and only supports x86_64 i.e. amd64 for linux, Apple M1 and intel for Mac and x86 and x86_64 for windows. There are the fully supported platforms along with architectures supported by playwright.

**Comment by fcollonval** [comment]
Thanks both for all the information. And thank you @mxschmitt for clarifying the documentation.

**Comment by kumaraditya303** [comment]
@conda-forge-admin ping team

**Comment by kumaraditya303** [comment]
@conda-forge-admin @conda-forge/help-python Ready for review

**Comment by fcollonval** [comment]
@kumaraditya303 you could try the gitter channel to get a reviewer feedback: https://gitter.im/conda-forge/conda-forge.github.io

File: recipes/playwright/meta.yaml
Comment (wolfv) [inline recipes/playwright/meta.yaml:18]: can you clarify why this is needed? is playwright shipping some binaries (for browsers?)

**Comment by wolfv** [comment]
hey, the recipe looks good to me! 
I don't understand why the binary relocation comments are needed and would like some clarification. Is chrom(ium) / firefox shipped within this bundle?

File: recipes/playwright/meta.yaml
Comment (kumaraditya303) [reply recipes/playwright/meta.yaml:18]: Playwright ships a driver for browsers to work

**Comment by kumaraditya303** [comment]
The binary relocation logic is required because the driver only works for newer glibc versions which is not present in the CI image hence we need to force it to skip it. On the systems which support include ubuntu 18 and 20 which both has that version

**Comment by wolfv** [comment]
Do you know which glibc version that is? You could add a constraint with `- __glibc >=2.17  # [linux]` to make use of the virtual packages in conda (and prevent installation on systems with too old glibc).

**Comment by kumaraditya303** [comment]
But playwright only supports Ubuntu 18 and 20 and I don't think I can constraint that

**Comment by kumaraditya303** [comment]
We have docs clearly stating that playwright requires Ubuntu 18 or Ubuntu 20 so there is not such need I think

**Comment by xhochy** [comment]
From CI:

> `INFO (playwright,lib/python3.7/site-packages/playwright/driver/node): `

Is this somehow shipping its own `node` binary? Or this is something else? We probably need to ship LICENSES for that, too.

> We have docs clearly stating that playwright requires Ubuntu 18 or Ubuntu 20 so there is not such need I think

conda installs are independent of the distribution. The analysis in CI also shows that you only depend on `glibc`  and `libstdc++`. Can you please find out which version and put that in the requirements as @wolfv stated.

**Comment by kumaraditya303** [comment]
Yes it ships its own node for running the browser driver and the LICENSE for it is included when the driver is downloaded.

**Comment by kumaraditya303** [comment]
Well it requires glibc version 2.17 or greater.

**Comment by xhochy** [comment]
> Yes it ships its own node for running the browser driver and the LICENSE for it is included when the driver is downloaded.

Is there any reason not to use conda-forge's `node` binary?

**Comment by kumaraditya303** [comment]
How can I add the glibc constraint where do I have to put - `__glibc >=2.17 # [linux]` ?

**Comment by kumaraditya303** [comment]
Using conda-forge would complicate the packaging as the driver is launched via a shell script on linux and mac and batch file on windows and with path hardcoded.

**Comment by kumaraditya303** [comment]
How can I add the glibc constraint where do I have to put - `__glibc >=2.17 # [linux]` ? @wolfv  @xhochy

**Comment by wolfv** [comment]
you can put it in run_constrained. 

E.g.

```
requirements:
   run:
     -  ... 
   run_constrained:
    - __glibc >=MYGLIBCVERSION  # [linux]
```

**Comment by wolfv** [comment]
@kumaraditya303 you can just use the node that is on the `PATH` (you don't need to hardcode anything) ... with conda this exactly should be super easy. Nodejs from conda-forge works really really well.

**Comment by kumaraditya303** [comment]
But that would mean we have to change the entire build chain for all the packages, fyi the driver is used for many other interfaces such as java and csharp and it would require changes across the packages, we don't want to complicate things now, we can see it later

**Comment by kumaraditya303** [comment]
@wolfv On CI 
```console
/home/conda/staged-recipes/build_artifacts/playwright_1619698706579/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/lib/python3.8/site-packages/playwright/driver/node: /lib64/libc.so.6: version `GLIBC_2.14' not found (required by /home/conda/staged-recipes/build_artifacts/playwright_1619698706579/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/lib/python3.8/site-packages/playwright/driver/node)

```


As far as I think what we are doing is useless as we are doing it just to pass the test on a linux distro which we don't support and I would like to ignore it and keep it simple as users of playwright would use it on Ubuntu 18 or 20.

**Comment by kumaraditya303** [comment]
@wolfv CI is green and would request you to merge the PR asap

**Review by xhochy — CHANGES_REQUESTED** [review]
Not ready for a merge, putting an explicit ❌ here.

This needs to either symlink to the conda-forge provided `node` or include its third-party notices.

Also it seems that the recipe is for a pre-release version. Can this be done for a release?

**Comment by kumaraditya303** [comment]
@xhochy I have already said that License is included in the driver package so why do you  want it twice any reason ?

**Comment by kumaraditya303** [comment]
@xhochy Why is pre-release versions not allowed in conda ?

**Comment by xhochy** [comment]
> @xhochy I have already said that License is included in the driver package so why do you want it twice any reason ?

It should be explicitly part of the package metadata to be correctly visible.

**Comment by xhochy** [comment]
> @xhochy Why, pre-release versions not allowed in conda ?

Every package on a channel is a release, there is no such thing as a pre-release. You can push the result to a different channel  but need to follow the docs at https://conda-forge.org/docs/maintainer/knowledge_base.html#pre-release-builds

**Comment by kumaraditya303** [comment]
We have packages on PyPI and there is no such constraint about adding third party LICENSE in metadata, only the LICENSE file is shipped that's all

**Comment by kumaraditya303** [comment]
@xhochy Third Party LICENSE is included now and the release version is now a stable release now can you merge this ?

**Comment by isuruf** [comment]
> As far as I think what we are doing is useless as we are doing it just to pass the test on a linux distro which we don't support and I would like to ignore it and keep it simple as users of playwright would use it on Ubuntu 18 or 20.

This is not the correct way to do this. Please see our docs on using a new sysroot.

**Comment by kumaraditya303** [comment]
We don't need that as our users wouldn't use that old distro as it is not supported. You can better ignore that.

**Comment by kumaraditya303** [comment]
@xhochy I have addressed the required changes and CI is green now you can merge this

**Comment by isuruf** [comment]
We can't ignore it. Please make the required changes or this PR will not be merged.

**Comment by kumaraditya303** [comment]
@isuruf I have marked that test to be not run on linux, Now are you okay with it ?

File: recipes/playwright/meta.yaml
Comment (kumaraditya303) [inline recipes/playwright/meta.yaml:45]: @isuruf

**Comment by isuruf** [comment]
Nope. That doesn't change anything.

**Comment by mxschmitt** [comment]
Hi, sorry for jumping in. I'm one of the maintainers of Playwright and try find there a solution to provide conda support for our Python users. Let's start maybe with something about [Playwright](https://github.com/microsoft/playwright) itself. Its a Node.js library which makes it possible to automate browsers (Chromium, Firefox, WebKit). It also has a functionality called a "driver" which makes it possible to add easy support for different languages like Python etc. which get shipped with a Node.js executable in the end.

To understand the problem better, the issue which is left, is that there is nothing in the meta-data of the package that a certain glibc version is required to use this package?

Playwright only supports Ubuntu 18.04 and 20.04 officially because the browsers are also custom compiled on these linux distributions. Any non LTS release is not supported and ends up mostly in not satisfied dependencies and won't get support on our issue tracker. (Windows + Mac are also supported)

cc @isuruf

**Comment by isuruf** [comment]
> What do you recommend there if we can't use the conda Node.js?

Please see our docs on using a new sysroot (for a newer glibc).

**Pull request closed without merging**
